### PR TITLE
Adds bias terms

### PR DESCRIPTION
**Pull request by rasmusbergpalm**
I've spent some time replicating the results of the original paper, and in my experience these bias terms are crucial.

Without:
image

With:
image

loss with and without
image

### CODE REVIEW

**Comment by lucidrains** [comment]
@rasmusbergpalm thank you! :D

**Comment by rasmusbergpalm** [comment]
You're welcome :) 

Here's my Slot Attention implementation that reproduced the results. I'm pretty confident it's the same as yours but having another pair of eyes can't hurt.

```
class SlotAttention(nn.Module):

    def __init__(self, input_size: int, hidden_size: int, slot_size: int, mlp_size: int, learn_slots=True):
        super().__init__()
        self.slot_size = slot_size
        self.hidden_size = hidden_size
        self.temp = 1 / t.sqrt(t.scalar_tensor(hidden_size))

        # Parameters
        self.learned_slots = learn_slots
        if learn_slots:
            self.slot_mean = nn.Parameter(t.zeros(slot_size), requires_grad=True)
            self.slot_logstd = nn.Parameter(t.ones(slot_size), requires_grad=True)
        self.k = nn.Linear(input_size, hidden_size, bias=True)
        self.v = nn.Linear(input_size, hidden_size, bias=True)
        self.q = nn.Linear(slot_size, hidden_size, bias=True)
        self.gru = nn.GRUCell(hidden_size, slot_size)
        self.mlp = nn.Sequential(
            nn.Linear(slot_size, mlp_size),
            nn.ReLU(),
            nn.Linear(mlp_size, slot_size)
        )
        self.ln_inputs = LayerNorm(input_size)
        self.ln_slots = LayerNorm(slot_size)
        self.ln_slots_pre_ff = LayerNorm(slot_size)

    def forward(self, inputs: t.Tensor, n_slots: int, n_attention_steps: int, eps=1e-8, slots: t.Tensor = None) -> t.Tensor:
        """

        :param inputs: (bs, n_inputs, input_dim)
        :param n_slots:
        :param n_attention_steps:
        :param slots:
        :param eps:
        :return: (bs, n_slots, slot_size)
        """

        bs = inputs.shape[0]

        assert slots is not None or self.learned_slots, "Initial slots must be learned or provided"
        if slots is None:
            slots = Normal(self.slot_mean, self.slot_logstd.exp()).rsample((bs, n_slots))  # (bs, n_slots, slot_size)

        inputs = self.ln_inputs(inputs)  # line 3

        k = self.k(inputs)  # (bs, n_inputs, hidden_size)
        v = self.v(inputs)  # (bs, n_inputs, hidden_size)
        for i in range(n_attention_steps):  # line 4
            slots_prev = slots  # line 5
            slots = self.ln_slots(slots)  # line 6

            # eq. 1, line 7 (bs, n_inputs, hidden_size) @ (bs, hidden_size, n_slots) = (bs, n_inputs, n_slots)
            M = self.temp * k @ self.q(slots).permute((0, 2, 1))  # (bs, n_inputs, n_slots)
            attn = nn.functional.softmax(M, dim=2) + eps  # (bs, n_inputs, n_slots)

            # eq. 2, line 8
            W = attn / attn.sum(dim=1, keepdim=True)  # (bs, n_inputs, n_slots)
            updates = W.permute((0, 2, 1)) @ v  # (bs, n_slots, n_inputs) @ (bs, n_inputs, hidden_size) = (bs, n_slots, hidden_size)

            # Line 9, applying the GRU
            updates = updates.reshape(bs * n_slots, self.hidden_size)
            slots_prev = slots_prev.reshape(bs * n_slots, self.slot_size)
            slots = self.gru(updates, hx=slots_prev)
            slots = slots.reshape(bs, n_slots, self.slot_size)

            # Line 10
            slots = slots + self.mlp(self.ln_slots_pre_ff(slots))

        return slots
```

Best,
Rasmus.

**Comment by DesertsP** [comment]
@rasmusbergpalm  Hello, could you please share your code of this experiment? I am a newbie in this field. Thanks.
